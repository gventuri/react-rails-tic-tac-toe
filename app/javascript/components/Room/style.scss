.row {
  .col {
    .cell {
      border-radius: 10px;
      border: 1px solid #999;
      background-color: #eee;
      width: 100%;
      padding-top: 100%; /* 1:1 Aspect Ratio */
      position: relative;
      background-position: center;
      background-repeat: no-repeat;

      &.cell-x {
        background-image: url("/assets/choices/x.png");
      }

      &.cell-o {
        background-image: url("/assets/choices/o.png");
      }
    }
  }
}
