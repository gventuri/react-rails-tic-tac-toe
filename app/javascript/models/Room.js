import axios from "axios";

class Room {
  static create({ slug, inviterCode, inviterStarts }) {
    return axios.post("/rooms", {
      slug,
      inviter_code: inviterCode,
      inviter_starts: inviterStarts,
    });
  }

  static join({ slug, challengerCode }) {
    return axios
      .post(`/rooms/${slug}/join`, {
        challenger_code: challengerCode,
      })
      .then((res) => res.data);
  }
}

export default Room;
