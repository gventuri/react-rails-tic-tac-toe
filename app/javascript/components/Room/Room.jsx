import React, { useState } from "react";
import { useParams } from "react-router-dom";
import { Form, InputGroup, Alert } from "react-bootstrap";
import { CopyToClipboard } from "react-copy-to-clipboard";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faClipboard } from "@fortawesome/free-solid-svg-icons";
import classNames from "classnames";
import PropTypes from "prop-types";
import "./style.scss";

const RoomUrl = () => {
  const { roomId } = useParams();
  const roomUrl = `http://localhost:3000/rooms/${roomId}`;
  const [copied, setCopied] = useState(false);

  return (
    <>
      <Form.Group controlId="validationCustomUsername">
        <Form.Label>
          Invite other players (copy link to the clipboard)
        </Form.Label>
        <InputGroup>
          <Form.Control type="text" value={roomUrl} disabled />
          <InputGroup.Text>
            <CopyToClipboard
              text={roomUrl}
              onCopy={() => {
                setCopied(true);
                setTimeout(() => setCopied(false), 3000);
              }}
            >
              <span className="cursor-pointer">
                <FontAwesomeIcon icon={faClipboard} />
              </span>
            </CopyToClipboard>
          </InputGroup.Text>
        </InputGroup>
      </Form.Group>

      <Alert show={copied} variant="success" dismissible>
        Copied to the clipboard
      </Alert>
    </>
  );
};

const Cell = ({ symbol, onChange }) => (
  <div className="col col-4 p-2">
    <div
      className={classNames("cell", {
        [`cell-${symbol}`]: symbol,
      })}
      onClick={() => onChange("x")}
      role="button"
      tabIndex="0"
    />
  </div>
);

const Room = ({ defaultCells }) => {
  const [cells, setCells] = useState(defaultCells);

  const updateCell = (key, value) => {
    setCells({
      ...cells,
      [key]: value,
    });
  };

  return (
    <>
      <RoomUrl />

      <div className="row">
        {Object.keys(cells).map((key) => (
          <Cell
            onChange={(newValue) => updateCell(key, newValue)}
            symbol={cells[key]}
            key={key}
          />
        ))}
      </div>
    </>
  );
};

Room.propTypes = {
  defaultCells: PropTypes.objectOf(PropTypes.any),
};

Room.defaultProps = {
  defaultCells: {
    0: null,
    1: null,
    2: null,
    3: null,
    4: null,
    5: null,
    6: null,
    7: null,
    8: null,
  },
};

export default Room;
