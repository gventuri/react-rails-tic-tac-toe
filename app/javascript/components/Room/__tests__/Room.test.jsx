import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import Room from "../Room";

afterEach(cleanup);

const INITIAL_ENTRIES = "/rooms/my-awesome-match";

it("should render room component", () => {
  render(
    <MemoryRouter initialEntries={[INITIAL_ENTRIES]}>
      <Room />
    </MemoryRouter>
  );
});

it("matches snapshots", () => {
  const room = renderer
    .create(
      <MemoryRouter initialEntries={[INITIAL_ENTRIES]}>
        <Room />
      </MemoryRouter>
    )
    .toJSON();
  expect(room).toMatchSnapshot();
});
