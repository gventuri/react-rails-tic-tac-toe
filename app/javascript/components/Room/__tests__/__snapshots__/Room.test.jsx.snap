// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
Array [
  <div
    className="form-group"
  >
    <label
      className="form-label"
      htmlFor="roomUrl"
    >
      Invite other players (copy link to the clipboard)
    </label>
    <div
      className="input-group"
    >
      <input
        className="form-control"
        disabled={true}
        id="roomUrl"
        type="text"
        value="http://localhost:3000/rooms/undefined"
      />
      <span
        className="input-group-text"
      >
        <span
          className="cursor-pointer"
          onClick={[Function]}
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-clipboard fa-w-12 "
            data-icon="clipboard"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M384 112v352c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V112c0-26.51 21.49-48 48-48h80c0-35.29 28.71-64 64-64s64 28.71 64 64h80c26.51 0 48 21.49 48 48zM192 40c-13.255 0-24 10.745-24 24s10.745 24 24 24 24-10.745 24-24-10.745-24-24-24m96 114v-20a6 6 0 0 0-6-6H102a6 6 0 0 0-6 6v20a6 6 0 0 0 6 6h180a6 6 0 0 0 6-6z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </span>
      </span>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
    <div
      className="col col-4 p-2"
    >
      <div
        className="cell"
        onClick={[Function]}
        role="button"
        tabIndex="0"
      />
    </div>
  </div>,
]
`;
